rfcs/0042-config-option.md

```diff
@@ -0,0 +1,301 @@
+---
+feature: config-option
+start-date: 2019-03-10
+author: Silvan Mosberger
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+A lot of NixOS options exist to specify single settings of configuration files. Along with such options come multiple disadvantages, such as having to synchronize with upstream changes, option bloat and more. An alternative is to provide options for passing the configuration as a string, such as the common `configFile` or `extraConfig` options, but this also comes with a set of disadvantages, including difficulty to override values and bad modularity.
+
+This RFC aims to solve these problems by encouraging NixOS module authors to provide an option for specifying the programs configuration file as a Nix value, providing a set of utility functions for writing these options conveniently, and updating the documentation to recommend this way of doing program configuration.
+
+# Motivation
+[motivation]: #motivation
+
+NixOS commonly has 2 models of specifying configuration for programs, each with their own set of problems. This RFC aims to solve all of them.
+
+## Single option for every setting
+
+Having a single option for every setting in the configuration file, this often gets combined with an `extraConfig` option to provide greater flexibility. Problems:
+
+- Coupling to upstream
+  - When upstream adds or removes settings, the NixOS module needs to be updated to reflect that.
+    - Upstream adds a setting: If the module has an `extraConfig` option people might set the new setting there. But if we ever add it as a NixOS option, we'll have trouble merging the values together with what the user already specified in `extraConfig`
+    - Upstream removes a setting (backwards incompatible): The NixOS module is straight up broken in nixpkgs until somebody fixes it, end users can't fix it themselves (unless the module provides a `configFile` option which can override the generated one). The same can also happen when the user uses overlays or `disabledModules` to cause a module/package version mismatch.
+  - The documentation of the upstream settings needs to be copied over to the NixOS options, which means it might get out of sync with changes
+  - The upstream defaults are being copied to the NixOS modules, so we need to also update the defaults whenever upstream changes them. This can be solved by using `nullOr` to allow for a `null` value to indicate that the upstream default shall be used, but that's not very nice.
```

Comment:
I've been debating this with myself for some time. In an ideal world I would prefer for the module to avoid using `null` to indicate "use the upstream default" and instead actually have the upstream default as the option default within the module.

The benefit being that the documentation shows the value that will be used and if the default changes upstream then the user will not be directly affected. Instead we could guard the new default behind a state version change allowing the user some agency.

That said, it seems to me that this is unfeasible in most cases unless the module maintainer is intimately aware of upstream changes.